### ISSUE TITLE
(PIV) Unknown error: index out of range

### ISSUE THREAD

**Issue by CaptAintHere** [Yubico/yubikey-manager-qt#256]
Hello all,

I'm not able to generate a new self-signed certificate with the RSA2048 algorithm on my Yubikey NEO anymore, I receive the following error : `Unknown error: index out of range`. If I select other algorithms such as RSA1024 et ECCP256 the certificate generation works, it is only the RSA2048 that doesn't work. Before having any issue there was only one certificate loaded in Slot 9a, generated with RSA2048. I just wanted to recreate a certificate in the same slot 9a, and now I can't create a certificate in any slot.

I've tried to completely reset PIV but it doesn't work either.
Does anybody have an idea on what is the issue here?

Here is more information about the environment :
OS : Windows 10
YubiKey : YubiKey NEO
YubiKey Manager v1.1.5

**Comment by fdennis** [comment]
Hi,
Thanks for the report!
Could you specify at what exact step the problem appears? Is it when you press "Generate" - "Self-signed certificate" - Choose "RSA2048" from the list. Is it at that point you cannot choose RSA2048?

**Comment by CaptAintHere** [comment]
Hi,
The error appears at the final step, when I press Generate, after choosing the subject name, the expiration date etc. There's a small delay between the moment I press Generate and the errors appears.
Let me know if debug logs could help!
Thanks

**Comment by fdennis** [comment]
Yes, please, post the debug log here.

**Comment by CaptAintHere** [comment]
Here is the debug log :
```
C:\Program Files\Yubico\YubiKey Manager>ykman.exe -l DEBUG piv generate-key --algorithm RSA2048 9a pubkey.pem
2021-02-10T09:28:46+0100 INFO [ykman.logging_setup.setup:58] Initialized logging for ykman version: 3.1.1
2021-02-10T09:28:46+0100 DEBUG [ykman.descriptor.Descriptor.open_device:86] transports: 0x4, self.mode.transports: 0x7
2021-02-10T09:28:46+0100 DEBUG [ykman.descriptor.open_device:91] Opening driver for serial: None, type: YUBIKEY.NEO, mode: OTP+FIDO+CCID
2021-02-10T09:28:46+0100 DEBUG [ykman.descriptor.open_device:94] Attempt 1 of 10
2021-02-10T09:28:46+0100 DEBUG [ykman.descriptor.open_device:97] Found driver: <ykman.driver_ccid.CCIDDriver object at 0x000001EBF325C310>, key_type: YUBIKEY.NEO, mode: OTP+FIDO+CCID
2021-02-10T09:28:46+0100 DEBUG [ykman.device.__init__:198] Read config from device...
2021-02-10T09:28:46+0100 DEBUG [ykman.device.__init__:213] Failed to read config from device
Traceback (most recent call last):
  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\device.py", line 199, in __init__
    config = DeviceConfig(driver.read_config())
  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\driver_ccid.py", line 197, in read_config
    raise NotSupportedError()
ykman.driver.NotSupportedError
2021-02-10T09:28:46+0100 DEBUG [ykman.device.__init__:119] Config data empty/missing
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040007a0000005272001'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'030500038f07060f00009000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'0001100000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'006ff6e19000'
2021-02-10T09:28:46+0100 DEBUG [ykman.device.__init__:227] Identified NEO
2021-02-10T09:28:46+0100 DEBUG [ykman.device.__init__:229] CCID available, probe capabilities...
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040007a0000005272001'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'030500038f07060f00009000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:213] Found applet: aid: AID.OTP , capability: OTP
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040008a0000006472f0001'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'5532465f56329000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:213] Found applet: aid: AID.U2F , capability: FIDO U2F
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040007a0000005271002'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'6d00'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:216] Missing applet: aid: AID.U2F_YUBICO , capability: FIDO U2F
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040005a000000308'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'61114f0600001000010079074f05a0000003089000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:213] Found applet: aid: AID.PIV , capability: PIV
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040006d27600012401'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'9000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:213] Found applet: aid: AID.OPGP , capability: OpenPGP
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040007a0000005272101'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'79030100017108708ab1bb0b9ca2349000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.probe_capabilities:213] Found applet: aid: AID.OATH , capability: OATH
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00a4040005a000000308'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'61114f0600001000010079074f05a0000003089000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00fd000000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'0100059000'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00cb3fff055c035fff00'
2021-02-10T09:28:46+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'530280009000'
Enter a management key [blank to use default key]:
2021-02-10T09:29:08+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'0087039b047c028000'
2021-02-10T09:29:08+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'7c0a800886467c5d897f79c29000'
2021-02-10T09:29:08+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'0087039b167c1480084cfc361a7c34079f8108bcb04788063a1537'
2021-02-10T09:29:08+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'7c0a8208f25070a76cdcf5fb9000'
2021-02-10T09:29:08+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'0047009a05ac03800107'
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'7f4982010981820100aa4985e03d58e3d63a96387f6c9a5bcfe8bb74570485ef5f754a491c5cb4e8862f3b603d4bfb5d1697985cdfe80fde48fa394fdb3e52b05ea223d82265a1a48cee47045de6254edb33dd61f5ea3c677384b3f1830939c5fba2b16f22e941fc04a8c861efed3ba87a88fa162b134446e2de935b576dc684caf58330fb369f1b047791b35b438723230a8bbd4dc14b079afe1ad4ab294ff6ac7eb5383a6f0568ba6adf3ff80cd583ce2ac3fb7edbf0ce8eca7c118bc73a774e816082878014b48b58baf334c289016a41280f27af0043b625ad283ee1f3fc717cb1c040458c192e367a2b34c756c78dd31af828402f8582ce2086956111'
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.send_apdu:228] SEND: b'00c0000000'
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.send_apdu:230] RECV: b'9000'
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.close:269] Close <ykman.driver_ccid.CCIDDriver object at 0x000001EBF325C310>
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.__del__:273] Destroy <ykman.driver_ccid.CCIDDriver object at 0x000001EBF325C310>
2021-02-10T09:29:29+0100 DEBUG [ykman.driver_ccid.close:269] Close <ykman.driver_ccid.CCIDDriver object at 0x000001EBF325C310>
"PyOtherSide error: Traceback (most recent call last):\n\n  File \"qrc:///py/cli.py\", line 12, in run\n    return main()\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\cli\\__main__.py\", line 273, in main\n    cli(obj={})\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 764, in __call__\n    return self.main(*args, **kwargs)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 717, in main\n    rv = self.invoke(ctx)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 1137, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 1137, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 956, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\core.py\", line 555, in invoke\n    return callback(*args, **kwargs)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\click\\decorators.py\", line 17, in new_func\n    return f(get_current_context(), *args, **kwargs)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\cli\\piv.py\", line 273, in generate_key\n    public_key = controller.generate_key(\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\piv.py\", line 794, in generate_key\n    key_data = Tlv.parse_dict(Tlv.unpack(0x7f49, resp))\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\util.py\", line 334, in parse_dict\n    return dict((tlv.tag, tlv.value) for tlv in cls.parse_list(data))\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\util.py\", line 328, in parse_list\n    tlv, data = cls.parse_from(data)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\util.py\", line 321, in parse_from\n    tlv = cls(data)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\util.py\", line 290, in __new__\n    ln, ln_ln = _tlv_parse_length(data, tag_ln)\n\n  File \"C:\\Program Files\\Yubico\\YubiKey Manager\\pymodules\\ykman\\util.py\", line 247, in _tlv_parse_length\n    ln = six.indexbytes(data, offs)\n\nIndexError: index out of range\n"
qml: Return value of PyObject call is NULL: Traceback (most recent call last):

  File "qrc:///py/cli.py", line 12, in run
    return main()

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\cli\__main__.py", line 273, in main
    cli(obj={})

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 717, in main
    rv = self.invoke(ctx)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\click\decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\cli\piv.py", line 273, in generate_key
    public_key = controller.generate_key(

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\piv.py", line 794, in generate_key
    key_data = Tlv.parse_dict(Tlv.unpack(0x7f49, resp))

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\util.py", line 334, in parse_dict
    return dict((tlv.tag, tlv.value) for tlv in cls.parse_list(data))

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\util.py", line 328, in parse_list
    tlv, data = cls.parse_from(data)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\util.py", line 321, in parse_from
    tlv = cls(data)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\util.py", line 290, in __new__
    ln, ln_ln = _tlv_parse_length(data, tag_ln)

  File "C:\Program Files\Yubico\YubiKey Manager\pymodules\ykman\util.py", line 247, in _tlv_parse_length
    ln = six.indexbytes(data, offs)

IndexError: index out of range
```

**Comment by fdennis** [comment]
It would be useful for us to know what firmware version and PIV version the key has. Could you run `ykman info` and `ykman piv info` and post the result here?

**Comment by CaptAintHere** [comment]
Sure, output of `ykman info` :
```
C:\Program Files\Yubico\YubiKey Manager>ykman.exe info
Device type: YubiKey NEO
Serial number: 7337697
Firmware version: 3.5.0
Enabled USB interfaces: OTP+FIDO+CCID

Applications
OTP Enabled
FIDO U2F Enabled
OpenPGP Enabled
PIV Enabled
OATH Enabled
FIDO2 Not available
```
And output of `ykman piv info` :
```
C:\Program Files\Yubico\YubiKey Manager>ykman.exe piv info
PIV version: 1.0.5
PIN tries remaining: 3
CHUID: 3019d4e739da739ced39ce739d836858210842108421c84210c3eb34106a7cd0d4992e45411cb01380be5fa6f3350832303330303130313e00fe00
CCC: f015a000000116ff02fbd0bfb22ac7f915b1627f519c8ef10121f20121f300f40100f50110f600f700fa00fb00fc00fd00fe00
```

**Comment by fdennis** [comment]
Have you tried this scenario with the same key on another computer? If not, could you try that? Just to check if this is a computer issue or not.

**Comment by CaptAintHere** [comment]
Hi @fdennis
Sorry for the long delay
I've just tried with another computer (same Windows version) and the same key, and it works... So it seems thst it's a computer issue...

**Comment by fdennis** [comment]
Yes, seems like it. I will close this issue then. By the way, we released a new version (1.2.0) yesterday. You could try giving that one a shot on the first computer you tried this on.
